odoo/cli/module.py

```diff
@@ -0,0 +1,172 @@
+            self.parser.error("Please provide a single database in the config file")
+        parsed_args.db_name = db_names[0]
+        if 'modules' in parsed_args and parsed_args.subcommand in ('install', 'upgrade'):
+            parsed_args.modules = self._get_module_names(parsed_args.modules)
+
+        match parsed_args.subcommand:
+            case 'install':
+                self._install(parsed_args)
+            case 'uninstall':
+                self._uninstall(parsed_args)
+            case 'upgrade':
+                self._upgrade(parsed_args)
+            case 'force-demo':
+                self._force_demo(parsed_args)
+
+    def _get_zip_path(self, path):
+        if (
+            (fullpath := Path(path).resolve())
+            and fullpath.exists()
+            and fullpath.suffix.lower() == '.zip'
+        ):
+            return fullpath
+        return None
+
+    def _get_module_names(self, module_names):
+        """ Get valid module names from disk before starting the Db environment """
+        initialize_sys_path()
+        module_names = set(module_names)
+        valid_module_names = {
+            module
+            for module in module_names
+            if get_module_path(module)
+            or self._get_zip_path(module)
+        }
+        if not valid_module_names:
+            _logger.warning("No valid module names found")
+        return valid_module_names
+
+    def _get_modules(self, env, module_names):
+        """ Get modules from the db, depending on the state """
+        Module = env['ir.module.module']
+        Module.update_list()
+        return Module.search([('name', 'in', module_names)])
+
+    @contextmanager
+    def _create_env_context(self, db_name):
+        with Registry.new(db_name).cursor() as cr:
+            yield odoo.api.Environment(cr, odoo.SUPERUSER_ID, {})
+
+    def _install(self, parsed_args):
+        with self._create_env_context(parsed_args.db_name) as env:
+
+            installable_modules = self._get_modules(env, parsed_args.modules)
+            parsed_args.modules = set(parsed_args.modules) - set(installable_modules.mapped("name"))
+            if installable_modules:
+                installable_modules.button_immediate_install()
+
+            importable_zipfiles = {
+                fullpath
+                for module in parsed_args.modules
+                if (fullpath := self._get_zip_path(module))
+            }
+            if 'imported' not in env['ir.module.module']._fields:
+                _logger.warning("Cannot import data modules unless the `base_import_module` module is installed")
+            else:
+                for importable_zipfile in importable_zipfiles:
+                    env['ir.module.module']._import_zipfile(importable_zipfile)
+
+    def _upgrade(self, parsed_args):
+        with self._create_env_context(parsed_args.db_name) as env:
+            if modules := self._get_modules(env, parsed_args.modules):
```

Comment:
Maybe this is a feature request, could you have an option that would upgrade all modules which have a newer version? Not sure this should be in this PR.

Reply:
I might have a brief look.

Reply:
Done, works well, useful.